src/middleware/extract.company.number.middleware.ts

```diff
@@ -0,0 +1,42 @@
+import { Request, Response, NextFunction } from "express";
+import { checkCompanyNumberFormatIsValidate } from "../utils/format/company.number.format";
+import { ContextKeys } from "../utils/constants/context.keys";
+import { getCompanyNumber } from "../utils/session";
+import { logger } from "../utils/logger";
+import { ValidateCompanyNumberFormat } from "../utils/validate/validate.company.number";
+
+declare module 'express-serve-static-core' {
+    interface Request {
+        companyNumber?: string;
+    }
+}
+
+export const extractCompanyNumberMiddleware = (req: Request, res: Response, next: NextFunction) => {
+    try {
+        const isSessionSet = typeof req.session !== "undefined";
+        const companyNumberParam = req.params.companyNumber;
+        let companyNumber: string = "";
+
+        if (companyNumberParam) {
+            checkCompanyNumberFormatIsValidate(companyNumberParam);
+            companyNumber = companyNumberParam;
+
+            if (isSessionSet && ValidateCompanyNumberFormat.isValid(companyNumber)) {
+                req.session!.setExtraData(ContextKeys.COMPANY_NUMBER, companyNumber);
+                logger.info(`Company Number set in session: ${companyNumber}`);
+            }
+        } else if (isSessionSet) {
+            let result = getCompanyNumber(req.session);
+            result = result instanceof Error ? "" : result;
+            companyNumber = result === "" ? companyNumber : result;
```

Comment:
This is quite messy, why not 
```suggestion
            let result = getCompanyNumber(req.session);
            if (! result instanceof Error) {
                companyNumber = result;
            }
```